lib/propolis/src/cpuid.rs

```diff
@@ -193,56 +197,223 @@ impl Specializer {
                             CpuidIdent::subleaf(leaf, 0),
                             CpuidValues {
                                 eax: 0x1,
                                 ebx: 0x2,
                                 ecx: 0x100,
-                                // TODO: populate with x2APIC ID
-                                edx: 0x0,
+                                edx: vcpuid,
                             },
                         );
                     } else {
-                        set.insert(
-                            CpuidIdent::subleaf(leaf, 0),
-                            CpuidValues {
-                                eax: 0x0,
-                                ebx: 0x1,
-                                ecx: 0x100,
-                                // TODO: populate with x2APIC ID
-                                edx: 0x0,
-                            },
-                        );
+                        // We notionally want to insert a leaf like
+                        // CpuidValues {
+                        //     eax: 0x0,
+                        //     ebx: 0x1,
+                        //     ecx: 0x100,
+                        //     edx: vcpuid,
+                        // }
+                        // here, but EAX=0 implies the leaf is invalid, rather
+                        // than the desired "shift x2APIC ID right by 0 to get
+                        // to the topology ID of processor cores"
+                        //
+                        // The question here is: what does this leaf look like
+                        // with hyperthreading disabled? Does this leaf return 0
+                        // in EAX implying that it is invalid, or is it 1, with
+                        // x2APIC IDs skipping every other entry for the
+                        // disabled SMT siblings? Whatever hardware does here is
+                        // least likely to surprise guest OSes.
+                        return Err(SpecializeError::IncompatibleTopology {
+                            leaf: 0xB,
+                            num_vcpu,
+                            why: Some("Leaf B.1 would have EAX=0"),
+                        });
+                    }
+                    // TODO: Not wholly clear if we should just set this to
+                    // `ceil(log2(num_vcpu))` (which should guarantee that the
+                    // VM is conceptually one socket) or set this to 7 or 8 like
+                    // a "normal" processor.
+                    //
+                    // Go with 8 for now and error if num_vcpu is above that:
+                    // you should trip over this quickly if you've bumped up
+                    // VM_MAXCPU and you may have hardware to compare against at
+                    // that point.
+                    if num_vcpu >= 256 {
+                        return Err(SpecializeError::IncompatibleTopology {
+                            leaf: 4,
```

Comment:
leaf B, not 4

Should TopoKind have some accessor like `leaf()` which emits a u32, so it could be passed in here, rather than typed in by hand?

Reply:
doubly awkward: `leaf`, which is used at various points in the match, _is_ the u32. the enum is also `#[repr(u32)]` which is why that value comes from a simple `*topo as u32`. i've struck the explicit numbers in these `IncompatibleTopology` errors in favor of just `leaf,`.